src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx

```diff
@@ -1,195 +1,91 @@
+                nodeType = firstLevelKey;
+            }
+        }
+    });
+
+    return {nodeType, type};
 }
 
-export function getFieldTemplate(type: string): FC<FieldProps> {
-    return getTemplate<FieldProps>(type, 'fields');
-}
-
-export function getElementTemplate(type: string): FC<ElementProps> {
-    return getTemplate<ElementProps>(type, 'elements');
-}
-
-export function getGroupTemplate(type: string): FC<GroupProps> {
-    return getTemplate<GroupProps>(type, 'groups');
-}
-
-export function getSectionTemplate(): FC<SectionProps> {
-    return getTemplate<SectionProps>('section', 'layouts', 'section');
-}
-
-export function getFormTemplate(): FC<FormProps> {
-    return getTemplate<FormProps>('form', 'layouts');
-}
-
-export function getFieldLabelTemplate(): FC<FieldLabelProps> {
-    return getTemplate('fieldLabel', 'layouts');
-}
-
-export function getFieldErrorTemplate(): FC<FieldErrorProps> {
-    return getTemplate('fieldError', 'layouts');
-}
-
-export function getHeaderTemplate(): FC<HeaderProps> {
-    return getTemplate('header', 'layouts');
-}
-
-export function getHeaderTitleTemplate(): FC<HeaderTitleProps> {
-    return getTemplate('headerTitle', 'layouts');
+/**
+ * This HOC will wrap a template component in our NodeWrapper and automatically figure out what nodeType and type to use as props.
+ *
+ * @unreleased
+ */
+export function withTemplateWrapper<TemplateProps>(
+    Template: FC<TemplateProps>,
+    htmlTag: keyof JSX.IntrinsicElements = 'div',
+    name?: string
+): FC<TemplateProps> {
+    const {nodeType, type} = findTemplateKeys(Template);
+
+    return (props: TemplateProps) => (
+        <NodeWrapper nodeType={nodeType} type={type} htmlTag={htmlTag} name={name}>
+            <Template {...(props as TemplateProps)} />
+        </NodeWrapper>
+    );
 }
 
-export function getHeaderDescriptionTemplate(): FC<HeaderDescriptionProps> {
-    return getTemplate('headerDescription', 'layouts');
-}
+/**
+ * A component version of withTemplateWrapper that uses the child component to determine the NodeWrapper props.
+ *
+ * @unreleased
+ */
+export function TemplateWrapper({children, htmlTag}: {children: JSX.Element; htmlTag?: keyof JSX.IntrinsicElements}) {
+    const {nodeType, type} = useMemo(() => findTemplateKeys(children.type), []);
 
-export function getGoalTemplate(): FC<GoalProps> {
-    return getTemplate('goal', 'layouts');
+    return (
+        <NodeWrapper nodeType={nodeType} type={type} htmlTag={htmlTag}>
+            {children}
+        </NodeWrapper>
+    );
 }
 
-function nodeIsFunctionalComponent(Node: unknown): Node is FC {
-    return typeof Node === 'function';
+/**
+ * A hook version of withTemplateWrapper
+ *
+ * @unreleased
+ */
+export function useTemplateWrapper<TemplateProps>(
+    Template: FC<TemplateProps>,
+    htmlTag: keyof JSX.IntrinsicElements = 'div',
+    name?: string
+): FC<TemplateProps> {
+    return useMemo(() => withTemplateWrapper(Template, htmlTag, name), [Template]);
```

Comment:
This `useMemo` is not likely to work reliably since the dependency is a function. But maybe it will since the templates don't change once loaded?